orttraining/orttraining/models/runner/training_runner.h

```diff
@@ -151,9 +152,18 @@ class TrainingRunner {
 
     int data_parallel_size = 1;
     int horizontal_parallel_size = 1;
-    int pipeline_stage_size = 1;
     // Enable gradient clipping.
     bool enable_grad_norm_clip=true;
+
+    // Pipeline configuration.
+
+    // model_paths[i] is the name of the i-th pipeline stage's forward model.
+    // The i-th stage is run by the i-th MPI rank. 
+    // If model_paths is not empty, model partition transformation may not be invoked.
+    VectorString pipeline_stage_paths;
+   // num_pipeline_stages > 1 means pipeline is enabled.
+   // num_pipeline_stages == 1 means pipeline is not enabled.
```

Comment:
For people don't know about pipeline, when they see num_pipeline_stages, they may set it to 0 to 'turn it off'. Shall we also allow 0 to be a legitimat input? We can internally reset it to 1 when we see it's a 0.

Also, consider adding a check for num_pipeline_stages internally, such that we can error out and exit early when we see a problem with user passed-in values.

Reply:
I am not a fan of having two definitions for one thing. It's convenient to someone but creates ambiguity to other people. Thanks.

Reply:
It is not having two definitions of one thing, it is actually exactly ambiguity caused by the name. Most generally, num_pipeline_stages = 0 means not using pipeline, as 0 pipeline stage is being used. And with some basic understanding of pipeline, num_pipeline_stages = 1 can also mean not using pipeline, as using 1 pipeline stage means the whole model is a whole and thus no model parallilism. For users who doesn't know pipeline at all, most likely they would set it to 0 as this is most intuitive, thus use the API wrong. 

If we want to make the definition specific, another option is to have a big switch which controls the on/off of the pipeline. We can have a boolean `use_pipeline`. This value will be default to false, in which case, all configuration to pipeline_context_ is ignored. When this value is specifically set to true, we start configure pipeline. 

---
In reply to: [419770968](https://github.com/microsoft/onnxruntime/pull/3770#discussion_r419770968) [](ancestors = 419770968)

Reply:
What about num_pipeline_stages = std::min(num_pipeline_stages, 1)

---
In reply to: [419791823](https://github.com/microsoft/onnxruntime/pull/3770#discussion_r419791823) [](ancestors = 419791823,419770968)

Reply:
I like this idea. It is more permissive to users' input and auto-correct if it is wrong, (which is one step ahead of checking the value). Maybe we can add a log letting user know we overwrite it if the change actually takes place?

---
In reply to: [419792899](https://github.com/microsoft/onnxruntime/pull/3770#discussion_r419792899) [](ancestors = 419792899,419791823,419770968)

Reply:
Let's consider a case, learning rate. What if user's python script has a bug which accidently revert "0.001" to "-0.001"? If we do "max(0, 0.00001)", user may get misleading result. Similarly, `num_pipeline_stages` has a default value 1, if user specifies 0, we should throw. Otherwise, if user accidently uses "-3" where the actual value is "3", they may still get a runnable pipeline which is far away from the original intention.

Reply:
The default value is set in main.cc in user code. This is not safe as some other users may write their program and set a different default value, or even not set it at all. I think a better practice is to assign a default value here as mentioned in the comment below. This will solve the 'user-not-set-it' problem. 

As for the definition of `num_pipeline_stages` and what value is considered to turn on/off pipeline, I am ok with having '1' as 'off', as long as we make it easy for user to use it right and hard to use it wrong. Currently the only validation of `num_pipeline_stages` is in external main.cc. We should also add check inside ORT before using `num_pipeline_stages` value. If we consider 0 as illegal input, let's validate it internally and error out early in this case.


---
In reply to: [419882786](https://github.com/microsoft/onnxruntime/pull/3770#discussion_r419882786) [](ancestors = 419882786)

Reply:
Will check in training_session.